File: #4/main.py
# Given an array of integers, find the first missing positive integer in linear time and constant space. In other words, find the lowest positive integer that does not exist in the array. The array can contain duplicates and negative numbers as well.

# For example, the input [3, 4, -1, 1] should give 2. The input [1, 2, 0] should give 3.

# You can modify the input array in-place.


# time O(n), space O(n)
def missingIntegerWithSpaceN(numbers):
	s = set(numbers)
	i = 1

	while True:
		if i not in s:
			return i
		i += 1


# time O(n), space O(1)
def rearrange(numbers):
	nbPositives = 0
	for i, nb in enumerate(numbers):
		if nb > 0:
			numbers[i], numbers[nbPositives] = numbers[nbPositives], numbers[i]
			nbPositives += 1
	return nbPositives


# time O(n), space O(1)
def missingInteger(numbers):
	end = rearrange(numbers)
	for i, nb in enumerate(numbers[:end]):
		x = abs(nb)
		if x <= end:
			numbers[x - 1] = -abs(numbers[x - 1])

	for i, nb in enumerate(numbers[:end]):
		if nb > 0:
			return i + 1

	return end + 1


print(f"missingInteger([1, 2, 3, 4, 5]) = {missingInteger([1, 2, 3, 4, 5])}")

print(f"missingInteger([3, -2, 1]) = {missingInteger([3, -2, 1])}")

print(f"missingInteger([1, 2, 0, 4, 5]) = {missingInteger([1, 2, 0, 4, 5])}")

print(f"missingInteger([1, 0, 3, 0, 5]) = {missingInteger([1, 0, 3, 0, 5])}")

File: #4/Main.java

/*
Given an array of integers, find the first missing positive integer in linear time and constant space. In other words, find the lowest positive integer that does not exist in the array. The array can contain duplicates and negative numbers as well.

For example, the input [3, 4, -1, 1] should give 2. The input [1, 2, 0] should give 3.

You can modify the input array in-place.
*/

public class Main {
	static int sort(int[] numbers) {
		int end = 0;
		for (int i = 0; i < numbers.length; i++) {
			if (numbers[i] > 0) {
				int temp = numbers[end];
				numbers[end++] = numbers[i];
				numbers[i] = temp;
			}
		}
		return end;
	}

	static int missingInteger(int[] numbers) {
		int end = sort(numbers);

		for (int i = 0; i < end; i++) {
			int abs = Math.abs(numbers[i]);
			if (abs <= end && numbers[abs - 1] > 0) {
				numbers[abs - 1] = -numbers[abs - 1];
			}
		}

		for (int i = 0; i < end; i++) {
			if (numbers[i] > 0) {
				return i + 1;
			}
		}

		return end + 1;
	}

	public static void main(String[] args) {
		System.out.printf("missingInteger([1, 2, 3, 4, 5]) = %d%n", missingInteger(new int[] { 1, 2, 3, 4, 5 }));
		System.out.printf("missingInteger([3, -2, 1]) = %d%n", missingInteger(new int[] { 3, -2, 1 }));
		System.out.printf("missingInteger([1, 2, 0, 4, 5]) = %d%n", missingInteger(new int[] { 1, 2, 0, 4, 5 }));
		System.out.printf("missingInteger([1, 0, 3, 0, 5]) = %d%n", missingInteger(new int[] { 1, 0, 3, 0, 5 }));
		System.out.printf("missingInteger([66, 420, 33, 1, 1337]) = %d%n",
				missingInteger(new int[] { 66, 420, 33, 1, 1337 }));
	}
}
